### PR TITLE
saleem-latif/MAYN-198: Private themes installation

### PR DESCRIPTION
**Pull request by mattdrayer**
@saleem-latif @fredsmith, FYI

### CODE REVIEW

File: playbooks/roles/edx-themes/defaults/main.yml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+themes_user: "{{ themes_service_name }}"
+themes_home: "{{ COMMON_CFG_DIR }}/{{ themes_service_name }}"
+
+THEMES_CODE_DIR: "{{ themes_home }}/{{ themes_service_name }}"
+
+THEMES_GIT_IDENTITY: !!null
+THEMES_GIT_PROTOCOL: "{{ COMMON_GIT_PROTOCOL }}"
+THEMES_GIT_MIRROR: "{{ COMMON_GIT_MIRROR }}"
+THEMES_GIT_PATH: "{{ COMMON_GIT_PATH }}"
+THEMES_REPO: "sample-themes.git"
```
Comment (fredsmith) [inline playbooks/roles/edx-themes/defaults/main.yml:23]: this repo contains no themes.

File: playbooks/roles/edx-themes/tasks/main.yml [hunk 2]
```diff
@@ -0,0 +1,113 @@
...
+
+- name: validate git protocol
+  fail: msg='THEMES_REPOS.PROTOCOL must be "https" or "ssh"'
+  when: (item.PROTOCOL != "https") and (item.PROTOCOL != "ssh") and THEMES_REPOS is defined
+  with_items: THEMES_REPOS
+  tags:
+    - install
+    - install:code
+
+- name: install read-only ssh key
```
Comment (fredsmith) [inline playbooks/roles/edx-themes/tasks/main.yml:79]: This key should be cleaned up after the build is finished.

File: playbooks/roles/edx-themes/defaults/main.yml (hunk 1 above)
Comment (mattdrayer) [inline playbooks/roles/edx-themes/defaults/main.yml:23]: That's correct -- no themes for now -- we're going to move 'red-theme' to it in the near future

File: playbooks/roles/edx-themes/tasks/main.yml [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+    - install:base
+
+# Assumes that the home directory has been created above.
+# In some cases(vagrant boxes) the home directory gets created
+# but does not have the current owner and group. In vagrant for
+# example we were seeing it defaulting it to `root` for both.
+# The item that is a blank string ("") ensures the ownership
+# of the home directory is always correct before proceeding.
+- name: create themes home dirs
+  file: >
```
Comment (fredsmith) [inline playbooks/roles/edx-themes/tasks/main.yml]: please use the correct syntax.  eg. 

```
file:
    path: "{{ themes_home }}/{{ item }}"
    state: directory
    owner: "{{ themes_user }}"
    group: "{{ common_web_group }}"
```

File: playbooks/roles/edx-themes/tasks/main.yml [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+- name: validate git protocol
+  fail: msg='THEMES_REPOS.PROTOCOL must be "https" or "ssh"'
+  when: (item.PROTOCOL != "https") and (item.PROTOCOL != "ssh") and THEMES_REPOS is defined
+  with_items: THEMES_REPOS
+  tags:
+    - install
+    - install:code
+
+- name: install read-only ssh key
+  copy: >
```
Comment (fredsmith) [inline playbooks/roles/edx-themes/tasks/main.yml]: syntax

File: playbooks/roles/edx-themes/tasks/main.yml [hunk 5]
```diff
@@ -0,0 +1,113 @@
...
+    content="{{ item.SSH_KEY }}" owner={{ themes_user }}
+    group={{ themes_user }} mode=0600
+  when: item.PROTOCOL == "ssh" and THEMES_REPOS is defined
+  with_items: THEMES_REPOS
+  tags:
+    - install
+    - install:code
+
+- name: checkout code over ssh
+  git_2_0_1: >
```
Comment (fredsmith) [inline playbooks/roles/edx-themes/tasks/main.yml]: syntax

File: playbooks/roles/edx-themes/tasks/main.yml [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+  sudo_user: "{{ themes_user }}"
+  register: code_checkout
+  when: item.PROTOCOL == "ssh" and THEMES_REPOS is defined
+  with_items: THEMES_REPOS
+  tags:
+    - install
+    - install:code
+
+- name: checkout code over https
+  git_2_0_1: >
```
Comment (fredsmith) [inline playbooks/roles/edx-themes/tasks/main.yml]: syntax

File: playbooks/roles/edx-themes/tasks/main.yml [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+# dependencies:
+#   - role: themes
+#     when do_setup_themes
+#
+
+# Generating an ssh key so service users can do a git
+# clone over ssh for public repositories without any
+# additional configuration
+- name: create application user
+  user: >
```
Comment (fredsmith) [inline playbooks/roles/edx-themes/tasks/main.yml]: syntax

File: playbooks/roles/ecommerce/tasks/main.yml [hunk 8]
```diff
@@ -21,6 +21,15 @@
 #
 #
 
+# Copy themes directories.
+# while synchronizing we need to make sure that directory path ends with a trailing slash,
+# otherwise synchronize will copy the src dir instead of the files/dirs included in the src directory
+- name: synchronize theme directories
+  command: >
```
Comment (fredsmith) [inline playbooks/roles/ecommerce/tasks/main.yml]: syntax

File: playbooks/roles/edx-themes/tasks/main.yml (hunk 2 above)
Comment (fredsmith) [inline playbooks/roles/edx-themes/tasks/main.yml:79]: Put this as the last task in this role:

```
- name: remove read-only ssh key
  file:
    dest: "{{ themes_home }}/.ssh/{{ item.REPO }}"
    state: absent
  when: THEMES_REPOS is defined
  with_items: THEMES_REPOS
  tags:
    - install
    - install:code
```

**Comment by feanil** [comment]
It looks like @fredsmith has caught all the issues I could spot as well so I think this is good to go once they are addressed.

**Comment by mattdrayer** [comment]
Excellent -- we'll take care of the things you've noted -- thanks!!!

File: playbooks/roles/edx-themes/tasks/main.yml (hunk 2 above)
Comment (mattdrayer) [inline playbooks/roles/edx-themes/tasks/main.yml:79]: Done!

**Comment by mattdrayer** [comment]
@fredsmith -- pushed a commit just now to address the issues you noted -- can you please check my edits to make sure everything is correct?

**Comment by fredsmith** [comment]
:+1: lgtm.  The configuration secure change is merged, so I think this is ready to build some AMIs.

**Comment by mattdrayer** [comment]
Nice!  I'll merge when the checks finish and kick off a new stage AMI build using the latest release tag (no changes to ecommerce version) so we can take a look.

**Comment by feanil** [comment]
👍

**Pull request merged**